### ISSUE TITLE
2.1.2 Content Negotiation

### ISSUE THREAD

**Issue by ysmartin** [w3c/w3c-waet#77]
**Current text:** cookies, authentication and session tracking have been subsumed under content negotiation.

**Suggestion:** several changes together:
- Keep these three features as top-level, distinct features, separate from content negotiation.
- Rewrite the content negotiation feature, as suggested below. This goes according to what is usually understood as content negotiation; nonetheless, the notes would point out the relation to those other features (cookies, etc.) The suggested writing solves also some technical inaccuracies. It combines elements from the previous version of the document and this version.
- Move content negotiation to the end of 2.1, so that it does not refer to other features which have not yet been presented (or is it better the other way around?).

**Suggested text:**

Content negotiation is a characteristic of the HTTP(S) protocol that enables web servers to provide different variants of a web resource located at a particular web address (URL/URI), depending on the demands of the web browser accessing it. For example, the server can provide the same web page in different languages, in mobile vs. desktop versions, or in other variants, depending on the features of the web browser and other context information sent to the web server. From the accessibility perspective, this implies that one variant may present accessibility problems, while another one may be fully accessible (for instance, when the page is requested in another language), despite the fact that both variants are located at the same address. 

To test these different variants of the web content, evaluation tool users need to be able to configure the content negotiation parameters to emulate different types of web browsers and set different preferences. These content negotiation parameters allow specifying the preferred language(s), preferred resource format(s) and browser features. [REF... to e.g. http://www.w3.org/Protocols/rfc2616/rfc2616-sec12.html and http://tools.ietf.org/html/rfc7231 ]

Note: Apart from content negotiation, web servers may tailor the content they serve, depending on the value of any field received from the client in the HTTP request. Servers typically take into account: the referring page (see crawling), cookies (see cookies), user credentials (see authentication) and other session maintenance mechanisms (see session). In any case, servers can in theory make the content vary depending on any other HTTP header field.

Note: This feature only refers to the selection of the content variant on the server side, based on the fields of the HTTP request. Web developers also apply other techniques to present different contents, depending on the browser and its features, which are applied by scripts running on the client side (see content rendering).

Note: The content retrieved or presented can also depend on more complex sequences of interactions between the browser and the server. Testing these would require the use of automation techniques (see automation)

(I'm not quite convinced of the specific writing of the notes above presented, but that would be a suitable way to mention other related features, which are distinct from content negotiation).

**Rationale:** 
[Specifications for content negotiation](http://www.w3.org/Protocols/rfc2616/rfc2616-sec12.html), related documentation of [servers](http://httpd.apache.org/docs/2.4/content-negotiation.html) and of [browsers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation) talk about several aspects to be negotiated. Besides, [Transparent Content Negotiation (TCN)](https://www.ietf.org/rfc/rfc2295.txt) is an advanced method, whose standardization did not reach the final stage, but it is supported by current servers. However, **none of these is directly related to cookies, authentication or session tracking mechanisms.** They rather refer to usually stateless, client preferences. It should also be noted that it is not the client which performs the selection, but the server, according to (or disregarding) the client preferences (client-based negotiation also exists, but it is rarer). Note that the client may show different preferred options for each preference at the same time, and they can be qualified with a priority level.

A technology-independent explanation of content negotiation can be found in [RFC2703](http://ftp.isi.edu/in-notes/rfc2703.txt) (by the way, the editor mentions that content negotiation can be used to negotiate accessible versions according to user's capabilities).

These are the main features that can be negotiated (together with the corresponding HTTP request/response header fields):
- Content type, media type or MIME (Accept / Content-Type)
- Character set (Accept-Charset / Content-Type), less used nowadays and not directly related to accessibility.
- Compression (Accept-Encoding / Content-Encoding), this one should not affect accessibility, as it only affects the transport layer.
- Language (Accept-Language / Content-Language)
- Transparent content negotiation: this is an advanced negotiation protocol that allows negotiating all the features listed above, plus an extended set of features (Negotiate, Accept-Features / TCN, Alternates) [Extended features](http://www.iana.org/assignments/media-feature-tags/media-feature-tags.xhtml) include e.g. the user agent's viewport size and resolution. This negotiation mechanism is seldom used.

Apart from these, a server can theoretically customize the content served, depending on the value of any other HTTP header field(s), or on any context property (some of these are hard to replicate e.g. time of the day, origin IP). Usually, servers make the content depend on:
- User agent features (User-Agent): the user agent sends a description of itself (vendor, version, rendering engine, capabilities, sometimes plugins) to the server in each HTTP request. The server may provide different responses to different user agents, through server-based browser detection, to offer browser-tailored versions, this is sometimes used, [although this is discouraged](https://developer.mozilla.org/en-US/docs/Browser_detection_using_the_user_agent). Other mechanisms to negotiate user agent features have been proposed, e.g. TCN mentioned above (Accept-Features), CC/PP exchange (Profile, Profile-Diff), but none of these have caught on. Note: [client-based techniques to detect user-agents and their features also exist](https://developer.mozilla.org/en-US/docs/Browser_Detection_and_Cross_Browser_Support), but they are not related to this feature, but to another in the last section.
- Referring page (Referer [sic]), the server serves different contents depending on the page from where the requested resource was referred to, that is, the previous page visited (when the requested resource is a page accessed through a link clicked somewhere else) or the containing HTML page (when the requested resource is an image, iframe, etc. embedded in a web page) Servers often used this feature to avoid hotlinking and cross-site request forgery attacks. Note that sending correct referer values can be necessary for proper site crawling, automation and session tracking.
- Other aspects mentioned at the beginning (cookies, authentication and session tracking) should also be given consideration  in the context of content customization in general, as the content served may depend on the user authenticated, etc., but this is not considered content negotiation, as above explained.

**_In summary**_, we have these features that can be negotiated and apply to accessibility evaluation: content language, content type, and user agent capabilities (for the lattest, proper negotiation mechanisms exist but are rarely supported; however, some server applications resort to "negoatiate" based on user agent identification chain). Plus, there are other related features which may make the content vary: referring page, cookies, authentication, session, and client-side user- agent detection and feature detection.
